portality/scripts/article_duplicates_report_remove.py

```diff
@@ -36,15 +37,20 @@ def duplicates_per_article(connection, delete, snapshot, owner=None, query_overr
         if duplicates:
             doi_dups = [d.id for d in duplicates.get('doi', [])]
             fulltext_dups = [d.id for d in duplicates.get('fulltext', [])]
-            doi_dups.sort(), fulltext_dups.sort()                            # Sort so we can visually compare the lists
+            fuzzy_dups = [d.id for d in duplicates.get('fuzzy', [])]
+            doi_dups.sort(), fulltext_dups.sort(), fuzzy_dups.sort()         # Sort so we can visually compare the lists
             print "\n{0}".format(article.id)
             if doi_dups:
                 print "\t{0} DOI duplicates: {1}".format(len(doi_dups), ", ".join(doi_dups))
             if fulltext_dups:
                 print "\t{0} fulltext duplicates: {1}".format(len(fulltext_dups), ", ".join(fulltext_dups))
+            if fuzzy_dups:
+                print "\t{0} fuzzy duplicates: {1}".format(len(fuzzy_dups), ", ".join(fuzzy_dups))
 
-            # Only consider duplicates that appear in both lists (we should be cautious with deletes)
-            set_of_duplicates = set(doi_dups).intersection(set(fulltext_dups))
+            # Only consider duplicates that appear in at least 2 lists (we should be cautious with deletes)
+            set_of_duplicates = reduce(set.union,
```

Comment:
Sorry, I have a little trouble following this through. Could you do me a favour and write out here what it's doing as a loop (instead of the `map`) and summary operation (instead of the `reduce`)?

Reply:
I'll just go a step further and explain it:

From inner to outer, we have:
```python
combinations([doi_dups, fulltext_dups, fuzzy_dups], 2)
```
Which gives us pairs of duplicate lists to compare. This ensures we only accept duplicates that appear in 2 of the lists. These are passed into:

```
map(lambda x: set(x[0]).intersection(set(x[1])), ...)
```
Which takes the paired duplicate lists, converts to set and calculates the intersection. Therefore we have a list of sets of the items that appear in 2 duplicate lists.

Finally, we want to get the full list of items that appear in 2 of the duplicate lists, so we combine them all with ```reduce(set.union, ...)``` using the union operation.

Reply:
So both the DOI and fulltext URL have to be the same for it to be considered a duplicate. Or fuzzy criteria + DOI or fulltext.

Could we put something to that effect in the function's description comment? (if I'm understanding it right)

The reason I asked for it to be written out more imperatively was partly to understand it more easily, but also partly to implicitly compare the readability of both approaches ;).